File: day2/solution_p2.clj
(require '[clojure.java.io :as io])
(require '[clojure.string :as string])

(defn parse [x y]
  (case x
    "forward" (list #(+ %1 (* %2 (Integer/parseInt y))) 0 (Integer/parseInt y))
    "down" (list (fn [x _] x) (Integer/parseInt y) 0)
    (list (fn [x _] x) (- (Integer/parseInt y)) 0)))

(defn parseMove [str]
  (apply parse (string/split str #" ")))

(defn sumTuples [acc next]
  (let [[a b c] acc
        [fn' x y] next]
    (list (fn' a b) (+ b x) (+ c y))))

(defn solution []
  (with-open [rdr (io/reader "../input.txt")]
    (->> (line-seq rdr)
         (map parseMove)
         (reduce sumTuples '(0 0 0))
         (apply #(* %1 %3)))))

File: day2/solution_p2.kt
package part2

import java.io.File

fun main() {
    println(solution())
}

fun String.parseMove(): Triple<(Int, Int) -> Int, Int, Int> =
    this.split(" ").let { s ->
        val value = s[1].toInt()
        when (s.first()) {
            "forward" -> Triple({ a, b -> b * value + a }, 0, value)
            "down" -> Triple({ a, _ -> a }, value, 0)
            else -> Triple({ a, _ -> a }, -value, 0)
        }
    }
//2006917119

fun solution(): Int {
    return File("../input.txt")
        .readLines()
        .map { it.parseMove() }
        .fold(Triple(0, 0, 0))
        { (a, b, c), (fn, x, y) -> Triple(fn(a, b), b + x, c + y) }
        .let { (f, _, s) ->
            f * s
        }
}
